### PR TITLE
Adding CratesApiKey validator

### LINKED ISSUE #529: create rule for https://crates.io/

### PR DESCRIPTION
**Pull request by eddynaka**
Fixes #529

### CODE REVIEW

File: Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security.Validators
+{
+    public class CratesApiKeyValidatorTests
+    {
+        [Fact]
+        public void CratesApiKeyValidator_Test()
+        {
+            string fingerprintText = "";
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs:23]: >"" [](http://example.com/codeflow?start=37&length=2)

use string.Empty for this pattern, just a style nit.

File: Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                return;
+            }
+
+            string message = null;
+            ResultLevelKind resultLevelKind = default;
+            var fingerprint = new Fingerprint(fingerprintText);
+            var keyValuePairs = new Dictionary<string, string>();
+
+            CratesApiKeyValidator.IsValidDynamic(ref fingerprint,
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs:34]: >IsValidDynamic [](http://example.com/codeflow?start=34&length=14)

What is this test validating? Just that we don't raise an exception?

File: Src/Plugins/Security/Security.SharedStrings.txt [hunk 3]
```diff
@@ -129,4 +129,5 @@
   $SEC101/044.NpmCredentialsUserPassword=(?is)(?:(?:(?:registry\s*=\s*https:\/\/(?P<host>[^\s]+)(?:$|\s))|(?:username\s*=\s*(?P<id>[^\s]+)(?:$|\s))|(?:_password\s*=\s*(?P<secret>[^\s]+)(?:$|\s)))(?:.{0,200})?){3}
   $SEC101/044.NpmCredentialsPassword=(?is)(?:(?:(?:registry\s*=\s*https:\/\/(?P<host>[^\s]+)(?:$|\s))|(?:_password\s*=\s*(?P<secret>[^\s]+)(?:$|\s)))(?:.{0,200})?){2}
   $SEC101/045.PostmanApiKey=\b(?P<secret>PMAK-[0-9a-z]{24}-[0-9a-z]{34})(?:[^0-9a-z]|$)
+  $SEC101/047.CratesApiKey=(?i)(?:[^c]|^)(?P<secret>cio[0-9a-z]{32})(?:[^0-9a-z]|$)
```
Comment (michaelcfanning) [inline Src/Plugins/Security/Security.SharedStrings.txt:132]: >(?:[^c]|^) [](http://example.com/codeflow?start=31&length=10)

we should push this pattern everywhere. in many other places we use '\b' here instead. Can we scrub for that?

Comment (eddynaka) [reply Src/Plugins/Security/Security.SharedStrings.txt:132]: I will make a new pr with this change :)

File: Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs (hunk 2 above)
Comment (eddynaka) [reply Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs]: removed.

File: Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security.Validators
+{
+    public class CratesApiKeyValidatorTests
+    {
+        [Fact]
+        public void DiscordCredentialsValidator_MockHttpTests()
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs:21]: >DiscordCredentialsValidator_MockHttpTests [](http://example.com/codeflow?start=20&length=41)

Why did you remove that other test? It was a good one.

Comment (eddynaka) [reply Src/Plugins/Tests.Security/Validators/SEC101_047.CratesApiKeyValidatorTests.cs:21]: It wasn't testing anything.
It was a blank test when I want to test a specific fingerprint.
